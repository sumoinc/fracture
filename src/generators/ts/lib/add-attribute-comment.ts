import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttributeComment = (f: TypeScriptSource, a: Attribute) => {
  f.line(`/**`);
  a.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
