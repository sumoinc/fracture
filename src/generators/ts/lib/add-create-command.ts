import { formatStringByNamingStrategy } from "../../../core/naming-strategy";
import { Shape } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addCreateCommand = (e: Shape) => {
  const command = `${e.fracture.namingStrategy.operations.commands.commandPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.commandSuffix}`;
  const commandInput = `${e.fracture.namingStrategy.operations.commands.inputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;
  const commandOutput = `${e.fracture.namingStrategy.operations.commands.outputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;

  const fileName = formatStringByNamingStrategy(
    command,
    e.fracture.namingStrategy.ts.file
  );
  const f = new TypeScriptSource(e.service, `/commands/${fileName}.ts`);

  const commandInputName = formatStringByNamingStrategy(
    commandInput,
    e.fracture.namingStrategy.model.shapeName
  );

  const commandOutputName = formatStringByNamingStrategy(
    commandOutput,
    e.fracture.namingStrategy.model.shapeName
  );

  f.open(`export interface ${commandInputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export interface ${commandOutputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export const handler = async (event: any) => {`);
  f.line("return true;");
  f.close(`};`);
  f.line("\n");
};
