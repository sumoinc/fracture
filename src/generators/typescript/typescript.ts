import { SourceCode } from "projen";
import { FractureComponent } from "../../core/component";
import { Fracture } from "../../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    // create index file
    const indexFile = typeScriptFile("index.ts");

    /**
     * Generate types for each entity.
     */
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      // generate helpful comments
      tsFile.line(`/**`);
      e.comment.forEach((c) => tsFile.line(` * ${c}`));
      tsFile.line(` */`);

      // build interface
      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        // generate helpful comments
        tsFile.line(`/**`);
        a.comment.forEach((c) => tsFile.line(` * ${c}`));
        tsFile.line(` */`);
        // console.log(a);

        // write type definition
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
