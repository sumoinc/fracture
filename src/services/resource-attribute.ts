import { paramCase } from "change-case";
import { Component } from "projen";
import { ValueOf } from "type-fest";
import { Resource } from "./resource";
import { Service } from "./service";

/**
 * Each Attribute has a type that is used to determine how we will construct other generated code.
 */
export const ResourceAttributeType = {
  /**
   *  A unique identifier for an object. This scalar is serialized like a String but isn't meant to be human-readable.
   *  Long format GUID
   */
  GUID: "GUID",
  /**
   *  In GraphQL:   A UTF-8 character sequence.
   *  In DynamoDB:  Maximum DynamoDB item size limit of 400 KB (DynamoDB limit)
   *                When used as DynamoDB Key:
   *                  For a simple primary key, the maximum length of the first attribute value (the partition key) is 2048 bytes.
   *                  For a composite primary key, the maximum length of the second attribute value (the sort key) is 1024 bytes.
   */
  STRING: "String",
  /**
   *  In GraphQL:   An integer value between -(231) and 232-1.
   *  In DynamoDB:  Stored as Number
   *                Numbers can be positive, negative, or zero. Numbers can have up to 38 digits of precision.
   *                Exceeding this results in an exception.
   *                Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
   *                Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
   */
  INT: "Int",
  /**
   *  In GraphQL:   An IEEE 754 floating point value.
   *  In DynamoDB:  Stored as Number
   *                Numbers can be positive, negative, or zero. Numbers can have up to 38 digits of precision.
   *                Exceeding this results in an exception.
   *                Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
   *                Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
   */
  FLOAT: "Float",
  /**
   *  In GraphQL:   A Boolean value, either true or false.
   */
  BOOLEAN: "Boolean",
  /*****************************************************************************
   *
   *  APPSYNC TYPES
   *
   *  Additional data types which AppSync supports.
   *
   ****************************************************************************/
  /**
   *  GraphQL:  An extended ISO 8601 date string in the format YYYY-MM-DD.
   *  DynamoDB: Stored as string
   */
  DATE: "Date",
  /**
   *  GraphQL:  An extended ISO 8601 time string in the format hh:mm:ss.sss.
   *  DynamoDB: Stored as string
   */
  TIME: "Time",
  /**
   *  GraphQL:  An extended ISO 8601 date and time string in the format YYYY-MM-DDThh:mm:ss.sssZ.
   *  DynamoDB: Stored as string
   */
  DATE_TIME: "DateTime",
  /**
   *  GraphQl:  An integer value representing the number of seconds before or after 1970-01-01-T00:00Z.
   *  DynamoDB: Stored as Number
   */
  TIMESTAMP: "Timestamp",
  /**
   *  GraphQL:  An email address in the format local-part@domain-part as defined by RFC 822.
   *  DynamoDB: Stored as string
   */
  EMAIL: "Email",
  /**
   *  GraphQL:  A JSON string. Any valid JSON construct is automatically parsed and loaded in the resolver mapping
   *            templates as maps, lists, or scalar values rather than as the literal input strings. Unquoted strings or
   *            otherwise invalid JSON result in a GraphQL validation error.
   *  DynamoDB: Stored as ???
   */
  JSON: "JSON",
  /**
   *  GraphQL:  A phone number. Phone numbers can contain either spaces or hyphens to separate digit groups. Phone
   *            numbers without a country code are assumed to be US/North American numbers adhering to the
   *            North American Numbering Plan (NANP).
   *  DynamoDB: Stored as string
   */
  PHONE: "Phone",
  /**
   *  GraphQL:  A URL as defined by RFC 1738. For example, https://www.amazon.com/dp/B000NZW3KC/ or
   *            mailto:example@example.com. URLs must contain a schema (http, mailto) and can't contain two forward
   *            slashes (//) in the path part.
   *  DynamoDB: Stored as string
   */
  URL: "URL",
  /**
   *  GraphQL:  A valid IPv4 or IPv6 address. IPv4 addresses are expected in quad-dotted notation (123.12.34.56).
   *            IPv6 addresses are expected in non-bracketed, colon-separated format (1a2b:3c4b::1234:4567). You can
   *            include an optional CIDR suffix (123.45.67.89/16) to indicate subnet mask.
   *  DynamoDB: Stored as string
   */
  IPADDRESS: "IPAddress",
  /*****************************************************************************
   *
   *  RELATIONSHIP TYPES
   *
   ****************************************************************************/
  ARRAY: "Array",
  MAP: "Map",
  /*****************************************************************************
   *
   *  FRACTURE TYPES
   *
   *  Special types used by Fracture. That are currently used primarily to store values calculated based on other
   *  attributes.
   *
   ****************************************************************************/
  COUNT: "Count",
  AVERAGE: "Average",
  SUM: "Sum",
} as const;

export const ResourceAttributeGenerator = {
  AUTO_INCREMENT: "Increment",
  GUID: "Guid",
  CURRENT_DATE_TIME_STAMP: "CurrentDateTimeStamp",
  TYPE: "Type",
  NONE: "None",
  VERSION_DATE_TIME_STAMP: "VersionDateTimeStamp",
  /**
   * Composed of other attributes
   */
  COMPOSITION: "Composition",
  /**
   * Genrated based on Tenant ID from SaaS Identity
   */
  TENANT: "Tenant",
} as const;

export const ManagementType = {
  /**
   * Users are intended to edit this value.
   */
  USER_MANAGED: "USER_MANAGED",
  /**
   * This value is intended to be automatically managed by the system.
   */
  SYSTEM_MANAGED: "SYSTEM_MANAGED",
} as const;

export const IdentifierType = {
  /**
   * This attribute is an identifier for this resource.
   */
  PRIMARY: "PRIMARY",
  /**
   * Not used as an identifier
   */
  NONE: "NONE",
} as const;

export const VisabilityType = {
  /**
   * Value can appear in API input / output.
   */
  USER_VISIBLE: "USER_VISIBLE",
  /**
   * System value, not shown or visible to user.
   */
  HIDDEN: "HIDDEN",
} as const;

export type ResourceAttributeOptions = {
  /**
   * The resource this attribute belongs to
   */
  readonly resource: Resource;

  /**
   * Full long name for this attribute.
   *
   * @example 'phone-number'
   */
  readonly name: string;

  /**
   * Brief name used when storing data to save space.
   *
   * @example 'pn'
   * @default AttributeOptions.name
   */
  readonly shortName?: string;

  /**
   * Is this attribute required?
   *
   * @default true
   */
  readonly required?: boolean;

  /**
   * Comment lines to add to the Resource.
   *
   * @default []
   */
  readonly comments?: Array<string>;

  /**
   * What is the type for tis attribute?
   *
   * @default ResourceAttributeType.STRING
   */
  readonly type?: ValueOf<typeof ResourceAttributeType> | Resource;

  /**
   * Type parameter for this structure attribute.
   *
   * @default any
   * @example 'T' for MyType<T> generic
   * @example 'Foo' for Array<Foo>
   */
  readonly typeParameter?: string | Resource;

  /**
   * Is this managed by the end user or the system?
   *
   * @default ManagementType.USER_MANAGED
   */
  readonly management?: ValueOf<typeof ManagementType>;

  /**
   * Is this value visible to the end user?
   *
   * @default VisabilityType.USER_VISIBLE
   */
  readonly visibility?: ValueOf<typeof VisabilityType>;

  /**
   * Is this value an identifier for this resource?
   *
   * @default IdentifierType.NONE
   */
  readonly identifier?: ValueOf<typeof IdentifierType>;

  /**
   * The separator to use when composing this attribute from other attributes.
   *
   * @default '#'
   */
  readonly compositionsSeperator?: string;

  /**
   * The generator to use for all operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  readonly generator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for create operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  readonly createGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for read operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  readonly readGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for update operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  readonly updateGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for delete operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  readonly deleteGenerator?: ValueOf<typeof ResourceAttributeGenerator>;
};

export class ResourceAttribute extends Component {
  /**
   * Returns an attribute by name, or undefined if it doesn't exist
   */
  public static byName(
    resource: Resource,
    name: string
  ): ResourceAttribute | undefined {
    const isDefined = (c: Component): c is ResourceAttribute =>
      c instanceof ResourceAttribute && c.name === name;
    return resource.attributes.find(isDefined);
  }

  /**
   * Returns an attribute by it's shortname, or undefined if it doesn't exist
   */
  public static byShortName(
    resource: Resource,
    shortName: string
  ): ResourceAttribute | undefined {
    const isDefined = (c: Component): c is ResourceAttribute =>
      c instanceof ResourceAttribute && c.shortName === shortName;
    return resource.attributes.find(isDefined);
  }

  /**
   * The resource this attribute belongs to
   */
  public readonly resource: Resource;

  /**
   * Full long name for this attribute.
   *
   * @example 'phone-number'
   */
  public readonly name: string;

  /**
   * Brief name used when storing data to save space.
   *
   * @example 'pn'
   * @default AttributeOptions.name
   */
  public readonly shortName: string;

  /**
   * Is this attribute required?
   *
   * @default true
   */
  public readonly required: boolean;

  /**
   * Comment lines to add to the Resource.
   *
   * @default []
   */
  public comments: Array<string>;

  /**
   * What is the type for tis attribute?
   *
   * @default ResourceAttributeType.STRING
   */
  public readonly type: ValueOf<typeof ResourceAttributeType> | string;

  /**
   * Type parameter for this structure attribute.
   *
   * @default undefined
   * @example 'T' for MyType<T> generic
   * @example 'Foo' for Array<Foo>
   */
  public readonly typeParameter?: string;

  /**
   * Is this managed by the end user or the system?
   *
   * @default ManagementType.USER_MANAGED
   */
  public readonly management: ValueOf<typeof ManagementType>;

  /**
   * Is this value visible to the end user?
   *
   * @default VisabilityType.USER_VISIBLE
   */
  public readonly visibility: ValueOf<typeof VisabilityType>;

  /**
   * Is this value an identifier for this resource?
   *
   * @default IdentifierType.NONE
   */
  public readonly identifier: ValueOf<typeof IdentifierType>;

  /**
   * Is this attribute value composed of other attribute values?
   * If so they are stored in this array.
   *
   * @default[]
   */
  public readonly compositionSources: Array<ResourceAttribute> = [];

  /**
   * The separator to use when composing this attribute from other attributes.
   *
   * @default: '#'
   */
  public readonly compositionsSeperator: string;

  /**
   * The generator to use for create operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  public readonly createGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for read operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  public readonly readGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for update operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  public readonly updateGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  /**
   * The generator to use for delete operations
   *
   * @default ResourceAttributeGenerator.NONE
   */
  public readonly deleteGenerator?: ValueOf<typeof ResourceAttributeGenerator>;

  constructor(
    public readonly project: Service,
    options: ResourceAttributeOptions
  ) {
    super(project);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.resource = options.resource;
    this.name = paramCase(options.name);
    this.shortName = options.shortName
      ? paramCase(options.shortName)
      : this.name;
    this.required = options.required ?? true;
    this.comments = options.comments ?? [];
    this.type =
      options.type instanceof Resource
        ? options.type.name
        : !options.type
        ? ResourceAttributeType.STRING
        : options.type;
    this.typeParameter =
      options.typeParameter instanceof Resource
        ? options.typeParameter.name
        : options.typeParameter;
    this.management = options.management ?? ManagementType.USER_MANAGED;
    this.visibility = options.visibility ?? VisabilityType.USER_VISIBLE;
    this.identifier = options.identifier ?? IdentifierType.NONE;
    this.compositionsSeperator = options.compositionsSeperator ?? "#";

    /***************************************************************************
     * Generators
     **************************************************************************/

    const defaultGenerator =
      options.generator ?? ResourceAttributeGenerator.NONE;
    this.createGenerator = options.createGenerator ?? defaultGenerator;
    this.readGenerator = options.readGenerator ?? defaultGenerator;
    this.updateGenerator = options.updateGenerator ?? defaultGenerator;
    this.deleteGenerator = options.deleteGenerator ?? defaultGenerator;

    /***************************************************************************
     * Add to Resource
     **************************************************************************/

    if (ResourceAttribute.byName(this.resource, this.name)) {
      throw new Error(
        `Resource "${this.resource.name}" already has an attribute named "${this.name}"`
      );
    }

    if (ResourceAttribute.byShortName(this.resource, this.shortName)) {
      throw new Error(
        `Resource "${this.resource.name}" already has an attribute with a shortname of "${this.shortName}"`
      );
    }

    this.resource.attributes.push(this);

    /***************************************************************************
     * Update primary resource structures
     **************************************************************************/

    // if visible, add to public data structure
    //if (this.visibility === VisabilityType.USER_VISIBLE) {
    this.resource.dataStructure.addAttribute({
      name: this.name,
      shortName: this.shortName,
      type: this.type,
      typeParameter: this.typeParameter,
      management: this.management,
      visibility: this.visibility,
      comments: this.comments,
      required: this.required,
    });
    //}

    return this;
  }

  public addCompositionSource(a: ResourceAttribute) {
    if (this.createGenerator !== ResourceAttributeGenerator.COMPOSITION) {
      throw new Error(
        "Cannot add composition source to non-composed attribute."
      );
    }
    this.compositionSources.push(a);
  }

  /***************************************************************************
   * Configuration export for this attribute
   **************************************************************************/

  public config(): Record<string, any> {
    return {
      name: this.name,
      shortName: this.shortName,
      required: this.required,
      comments: this.comments,
      type: this.type,
      typeParameter: this.typeParameter,
      management: this.management,
      visibility: this.visibility,
      identifier: this.identifier,
      compositionSources: this.compositionSources.map((a) => a.name),
      compositionsSeperator: this.compositionsSeperator,
      createGenerator: this.createGenerator,
      readGenerator: this.readGenerator,
      updateGenerator: this.updateGenerator,
      deleteGenerator: this.deleteGenerator,
    };
  }
}
