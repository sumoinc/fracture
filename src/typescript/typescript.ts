import { SourceCode } from "projen";
import { FractureComponent } from "../core/component";
import { Fracture } from "../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    const indexFile = typeScriptFile("index.ts");
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
