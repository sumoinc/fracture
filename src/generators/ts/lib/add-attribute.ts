import { addAttributeComment } from "./add-attribute-comment";
import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttribute = (f: TypeScriptSource, a: Attribute) => {
  addAttributeComment(f, a);

  const q = a.isRequired ? "" : "?";
  f.line(`${a.name}${q}: ${a.typeScriptType};`);
};
