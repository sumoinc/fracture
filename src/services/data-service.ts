import { Project } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { Resource, ResourceOptions } from "./resource";
import { Service, ServiceOptions } from "./service";
import { Structure, StructureOptions } from "./structure";
import { StructureAttributeType } from "./structure-attribute";
import { DynamoTable, DynamoTableOptions } from "../dynamodb";
import { TypescriptStrategy } from "../generators/ts/strategy";

export interface DataServiceOptions extends ServiceOptions {
  /**
   * Options for the dynamo table.
   *
   * @default - fracture defaults
   */
  readonly dynamoTableOptions?: DynamoTableOptions;
  /**
   * Options for resources to add when initializing the service.
   */
  readonly resourceOptions?: Array<ResourceOptions>;
  /**
   * Typescript strategy options
   *
   * @default fracture defaults
   */
  //typescriptStrategyOptions?: TypescriptStrategyOptions;

  // srcDir?: string;
  /**
   * Logging options
   * @default LogLevel.INFO
   */
  //logging?: LoggerOptions;
  /**
   * Versioned.
   * @default false
   */
  //isVersioned?: boolean;
  /**
   * The naming strategy to use for generated code.
   */
  // namingStrategy?: NamingStrategy;
  /**
   * The audit strategy to use for generated code.
   */
  //auditStrategy?: AuditStrategy;
}

export class DataService extends Service {
  /**
   * Returns a service by name, or undefined if it doesn't exist
   */
  public static byName(
    project: TypeScriptProject,
    name: string
  ): DataService | undefined {
    const isDefined = (c: Project): c is DataService =>
      c instanceof DataService && c.name === name;
    return project.subprojects.find(isDefined);
  }

  /**
   * Returns all services
   */
  public static all(project: TypeScriptProject): DataService[] {
    const isDefined = (c: Project): c is DataService =>
      c instanceof DataService;
    return project.subprojects.filter(isDefined);
  }

  /**
   * A name for the service.
   */
  public readonly name: string;
  /**
   * Defines typescript naming conventions for this service.
   */
  public readonly typescriptStrategy: TypescriptStrategy;
  /**
   * All resourcers in this service.
   */
  //public resources: Array<Resource> = [];
  /**
   * Table defnition for the table that supports this service.
   */
  // public readonly dynamoTable: DynamoTable;
  /**
   * All structures for this service.
   */
  // public structures: Array<Structure> = [];
  public readonly errorStructure: Structure;
  public readonly responseStructure: Structure;
  public readonly listResponseStructure: Structure;
  /**
   * Output directory for CDK artifacts (cdk.out)
   */
  //public readonly cdkOutBuildDir: string;
  /**
   * Where deployable artifacts are stored in pipeline runs
   */
  // public readonly cdkOutDistDir: string;

  constructor(options: DataServiceOptions) {
    super(options);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = options.name;
    this.typescriptStrategy = new TypescriptStrategy(this);

    /***************************************************************************
     * Dynamo Configuration
     **************************************************************************/

    if (options.dynamoTableOptions) {
      new DynamoTable(this, options.dynamoTableOptions);
    }

    /***************************************************************************
     * Initialize standard default structure shapes
     **************************************************************************/

    this.errorStructure = this.addStructure({
      name: `error`,
      attributeOptions: [
        {
          name: `code`,
          type: StructureAttributeType.INT,
        },
        {
          name: `source`,
        },
        {
          name: `message`,
        },
        {
          name: `detail`,
        },
      ],
    });

    this.responseStructure = this.addStructure({
      name: `response`,
      typeParameter: `T`,
      attributeOptions: [
        {
          name: `data`,
          type: StructureAttributeType.CUSTOM,
          typeParameter: `T`,
          required: false,
        },
        {
          name: `errors`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `error`,
        },
        {
          name: `status`,
          type: StructureAttributeType.INT,
        },
      ],
    });

    this.listResponseStructure = this.addStructure({
      name: `list-response`,
      typeParameter: `T`,
      attributeOptions: [
        {
          name: `data`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `T`,
          required: false,
        },
        {
          name: `errors`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `error`,
        },
        {
          name: `status`,
          type: StructureAttributeType.INT,
        },
      ],
    });

    /***************************************************************************
     * Resources based on options
     **************************************************************************/

    if (options.resourceOptions) {
      options.resourceOptions.forEach((resourceOption) => {
        this.addResource(resourceOption);
      });
    }
  }

  public addResource(options: ResourceOptions) {
    return new Resource(this, options);
  }

  public addStructure(options: StructureOptions) {
    return new Structure(this, options);
  }
}
