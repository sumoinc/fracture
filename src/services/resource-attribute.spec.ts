import { TypeScriptProject } from "projen/lib/typescript";
import { DataService } from "./data-service";
import { ResourceAttribute } from "./resource-attribute";

test("Smoke test", () => {
  const service = testService();
  const att = new ResourceAttribute(service, { name: "bar" });
  expect(att).toBeTruthy();
});
const testService = () => {
  return new DataService({
    parent: new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    }),
    name: "foo",
  });
};
