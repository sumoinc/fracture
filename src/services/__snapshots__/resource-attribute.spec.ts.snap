// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Attribute Relationships 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

export type Error = {
    code: number;
    source: string;
    message: string;
    detail: string;
};

export type Request<T> = {
    input?: T;
};

export type Response<T> = {
    data?: T;
    errors: Array<Error>;
    status: number;
};

export type ListRequest<T> = {
    /**
     * A search string to filter the list by.
     * Will match starting string in index text. Case insensitive.
    */
    indexTerm: string;
    nextToken: string;
};

export type ListResponse<T> = {
    data?: Array<T>;
    errors: Array<Error>;
    status: number;
};

/**
 * Public facing data structure for this record.
*/
export type Address = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    street: string;
    city: string;
};

/**
 * Private data structure for this record.
 * Used internally in Dynamo, SQS and other native services.
*/
export type AddressData = {
    /**
     * Partition Key for this record.
    */
    pk: string;
    /**
     * Sort Key for this record.
    */
    sk: string;
    /**
     * Lookup value for this record.
    */
    idx: string;
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    t: string;
    /**
     * Version for record.
    */
    v: string;
    /**
     * The timestamp representing when this record was created.
    */
    ct: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    mt: string;
    /**
     * The timestamp representing when this record was marked as deleted.
    */
    dt: string;
    street: string;
    city: string;
};

export type CreateAddressInput = {};

export type CreateAddressOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    street: string;
    city: string;
};

export type GetAddressInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type GetAddressOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    street: string;
    city: string;
};

export type UpdateAddressInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type UpdateAddressOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    street: string;
    city: string;
};

export type DeleteAddressInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type DeleteAddressOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    street: string;
    city: string;
};

/**
 * Public facing data structure for this record.
*/
export type User = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};

/**
 * Private data structure for this record.
 * Used internally in Dynamo, SQS and other native services.
*/
export type UserData = {
    /**
     * Partition Key for this record.
    */
    pk: string;
    /**
     * Sort Key for this record.
    */
    sk: string;
    /**
     * Lookup value for this record.
    */
    idx: string;
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    t: string;
    /**
     * Version for record.
    */
    v: string;
    /**
     * The timestamp representing when this record was created.
    */
    ct: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    mt: string;
    /**
     * The timestamp representing when this record was marked as deleted.
    */
    dt: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};

export type CreateUserInput = {};

export type CreateUserOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};

export type GetUserInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type GetUserOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};

export type UpdateUserInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type UpdateUserOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};

export type DeleteUserInput = {
    /**
     * Identifier for this record.
    */
    id: string;
};

export type DeleteUserOutput = {
    /**
     * Identifier for this record.
    */
    id: string;
    /**
     * Type of record.
    */
    type: string;
    /**
     * Version for record.
    */
    version: string;
    /**
     * The timestamp representing when this record was created.
    */
    createdTimestamp: string;
    /**
     * The timestamp representing when this record was last modified.
    */
    modifiedTimestamp: string;
    firstName: string;
    lastName: string;
    primaryAddress: Address;
    addresses: Array<Address>;
    addressMap: Record<string, Address>;
};
"
`;
